API definition of DisruptionBudget

mqliang opened the pull request:
Add API definition of DisruptionBudget so that it would not block the implementation of rescheduler/preemption/drain feature.

Just send for reviewing. If this definition looks good, I will update the auto generation files. 

@HaiyangDING @jiangyaoguo @davidopp @mml

davidopp commented:
ref/ #12611

davidopp commented:
ping @mml

mml commented on pkg/api/types.go:
```diff
@@ -1407,6 +1409,69 @@ type ReplicationControllerList struct {
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
```
I think that users generally want to specify MinAvailable.  They might seem like the same thing, but they're not.

Suppose a pod is evicted and its replacement has been sent to the kubelet, but it's not yet in ready state.  If MinReadySeconds haven't yet elapsed, one view is that the new pod isn't *yet* unavailable.

OTOH, no one would say the pod is available.  It will be less confusing to specify MinAvailable.

mqliang replied:
SGTM.

mml commented on pkg/api/types.go:
```diff
@@ -1407,6 +1409,69 @@ type ReplicationControllerList struct {
+	// ObservedDisruptionTokenNum if the observed number of DisruptionToken
+	ObservedDisruptionTokenNum inf.Dec `json:"observedDisruptionTokenNum,omitempty"`
+	// Duration is the time unit of disruption rate the collection of pods is experiencing.
+	Duration time.Duration `json:"duration,omitempty"`
```
How does this Duration differ from the one in RateLimitSpec?

mqliang replied:
No difference, so I think we could remove it.

mml commented on pkg/api/types.go:
```diff
@@ -1407,6 +1409,69 @@ type ReplicationControllerList struct {
+type DisruptionBudgetSpec struct {
+	// RateLimit is a rate limit on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimit *RateLimitSpec `json:"ratelimit,omitempty"`
```
This type allows only one rate limit.  At Google, we have found it useful to express guarantees over both a 30m and a 1d window.  E.g. the daily budget may allow for 100 disruptions in a day, but in any 30m window we can only cause 7.

I suggest we create the type such that a user can specify a list of rate limits, but also say that we only define what will happen if the user specifies exactly 1.  This lets us add support for multiple rate limits later without changing semantics on any previously supported specs.

@bgrant0607 can you verify this would follow all the API rules?

mqliang replied:
A list of rate limits SGTM.

mml replied:
> I suggest we create the type such that a user can specify a list of rate limits, but also say that we only define what will happen if the user specifies exactly 1. This lets us add support for multiple rate limits later without changing semantics on any previously supported specs.

I spoke with Brian and he confirmed that this is OK, with a few additions:

1. We should document how we think the API will be extended in the future.
2. We should reject budgets with >1 rate limit at the API server.
3. We should make it clear that other clients should properly deal with any number of rate limits rather than expecting to always see 1 rate limit.

mqliang replied:
@mml 
> 
2. We should reject budgets with >1 rate limit at the API server.
3. We should make it clear that other clients should properly deal with any number of rate limits rather than expecting to always see 1 rate limit.

I am a little confused.  I think API server could handle the multi-ratelimit situation gracefully: the "evict" subresource will ensure that a disruption could only happen when it meet ALL the ratelimits.
So, why should we reject budgets with >1 rate limit at the API server?

mml commented on pkg/api/types.go:
```diff
@@ -1407,6 +1409,69 @@ type ReplicationControllerList struct {
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	MinReadySeconds int `json:"minReadySeconds,omitempty"`
```
I understand why this is here, but I worry that the conjunction of this, activeDeadlineSeconds in the pod spec, the other assorted timeouts in the pod probe spec, and MinReadySeconds in deployment is confusing.

It would be nice if we could reduce the number of places the user had to configure these related timeouts.  Any ideas?

mqliang replied:
remove this in flavour of activeDeadlineSeconds in the pod spec? Or make this equal to activeDeadlineSeconds by default?

mqliang commented:
@mml BTW, I am not a native english speaker, any comments about the godocs/user documents are welcomed.

mml commented:
@mqliang I'll review the docs/comments.  In the meantime, are you planning to revise the PR with the edits we agreed upon?

mqliang commented:
@mml 
>  In the meantime, are you planning to revise the PR with the edits we agreed upon?

Will rework soon.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+type DisruptionToken struct {
+	// ServiceOrJobRef is reference to the Service or Job
+	ServiceOrJobRef LocalObjectReference `json:"serviceOrJobRef,omitempty"`
+}
```
This is too restrictive.

bgrant0607 replied:
Why not an independent api with a label selector?

davidopp replied:
For the first version, I think we should just have an (optional) pointer from each Pod to its DisruptionBudget. To address @bgrant0607's comment about v1 API, we could make it an annotation on Pod instead of a field of PodSpec. So in sum, you would
* remove the field you added to ServiceSpec
* remove the field you added to JobSpec
* remove this field in DisruptionToken
And then when we actually use the feature (e.g. user creates a Pod they want protected, or node drain tool is considering draining a machine), we would use the Pod annotation.

I think that for the second version, for Job and Service we might want to inline the DisruptionBudgetSpec and have the controller create a corresponding DisruptionBudget object (which it could reflect back in the JobStatus/ServiceStatus), rather than having the user specify a reference to a DisruptionBudget in the JobSpec/ServiceSpec.

davidopp replied:
Also, shouldn't the DisruptionToken point back to the Disruption object that it is a token for? In other words, I'm not sure it ever made sense for this field to point to Service/Job -- it should point to the Disruption object, which represents the token bucket it lives in. But why does this object exist at all? It seems the RateLimitStatus can just store a fixed-size array of timestamps, each one representing a "token" that will get deleted at timestamp + RateLimitSpec.duration.

mqliang replied:
>  But why does this object exist at all? It seems the RateLimitStatus can just store a fixed-size array of timestamps, each one representing a "token" that will get deleted at timestamp + RateLimitSpec.duration.

One reason of having DisruptionToken is we could use the TTL functionality of etcd to auto delete it when it expired. Storing a fixed-size array of timestamps in RateLimitStatus sounds good to me.

bgrant0607 replied:
Doesn't have to be the TTL functionality of etcd. We're going to have a garbage collector.

bgrant0607 replied:
`ServiceOrJobRef` is mixing concepts. Please don't do that.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits []RateLimitSpec `json:"ratelimits,omitempty"`
+	// The minium number of pods that must be available simultaneously.
```
rateLimits

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+	// The current disruption rate that the collection of pods is experiencing.
+	// This data may be out of date by some window of time.
+	RateLimitStatuses []RateLimitStatus `json:"RateLimitStatuses,omitempty"`
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
```
rateLimitStatuses

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1489,6 +1556,9 @@ type ServiceSpec struct {
 	Selector map[string]string `json:"selector"`
 
+	// DisruptionBudget defines the max disruption that can be caused to the collection of pods the Service matching
+	DisruptionBudget *DisruptionBudget `json:"disruptionBudget,omitempty"`
```
I'm not sure I like the reference approach for this

mml replied:
@bgrant0607 what specifically do you object to?

bgrant0607 replied:
DisruptionBudget is in Service and Job, which is like apples and oranges.

Additionally, this is ServiceSpec, but DisruptionBudget contains both spec and status.

bgrant0607 replied:
Also, don't add this to the v1 API. See @davidopp's comments about using an annotation.

mqliang replied:
@davidopp suggest to use Pod annotation. Why shouldn't we use Service and Job annotation? IMO, DisrutionBudget applies to a collection of Pods, so it should put in Service and Job(in their annotation in the first version).

bgrant0607 commented:
Let's find a way for this to not be in the v1 API

bgrant0607 commented:
Have you thought about how this should relate to forgiveness and tolerations?

mqliang commented:
@bgrant0607 
> Have you thought about how this should relate to forgiveness and tolerations?

There seems no strong relationship between DisruptionBudget and Forgiveness: 

* DisruptionBudget goes in Service/Job; Forgiveness goes in Pod; 
* DisruptionBudget is used to limit disruptions due to multiple causes (node drain, preemption, rescheduling etc.); Forgiveness is used to control over pod durability.

The only common point shared by DisruptionBudget and Forgiveness is that they both have threshold(ratelimit etc), and they can both use the "evict" subresource. So, if we want implement Forgiveness at the same time, I think we should:

* We will have two kind of Token: DisruptionToken and ForgivenessToken, used to limit the rate of pod disruption and forgiveness, respectively.
* Implement the "evict" subresource
* Move the related logic in gc_controller.go and node_controller.go to a forgive_controller.go: node_controller should just be responsible for monitoring the health status of node, forgive_controller will evict the pod from dead/unreachable nodes using the "evict" subresource, according to ForgivenessPolicy. 

Ok, It seems like it might be reasonable to implement DisruptionBudget/Forgiveness/Evict at the same time. Maybe I should send a proposal to describe the API definition of DisruptionBudget/Forgiveness and related work first, WDTK?

davidopp commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	MinReadySeconds int `json:"minReadySeconds,omitempty"`
```
Do we need this option in the first version?

davidopp commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits []RateLimitSpec `json:"ratelimits,omitempty"`
```
I'd prefer to not use "Spec" or "Status" in the type name for anything that is not an API resource. But if we're already violating that preference elsewhere, then ignore my comment.

bgrant0607 replied:
+1

mqliang replied:
How about: 
1) remove DisruptionBudget
2) rename DisruptionBudgetSpec as DisruptionBudgetStrategy, and put it in Service and Job(use Pod annotation in the first version). 
3) Put RateLimitStatuses, AvailableReplicas, UnavailableReplicas, etc in ServiceStatus and JobStatus, directly. Do not use nested definition.

davidopp commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+
+// RateLimitSpec describes the rate limit on disruptions (preemption and other evictions) across the corresponding
+// collection of pods
+type RateLimitSpec struct {
```
Just to make sure I understand, the meaning of the parameters is
* RateLimitSpec.Capacity: the size of the bucket
* RateLimitSpec.Duration: how long a token lives in the bucket, from insertion until deletion
* DeletionTimestamp associated with DisruptionToken: insertion timestamp + RateLimitSpec.Duration

Is that correct?

mqliang replied:
Yes, your understanding is right.

davidopp commented:
I think we should definitely not try to implement forgiveness, tolerations, and DisruptionBudget at the same time. It's worth thinking about them but I don't want us to get stuck over-planning, especially since DisruptionBudget is considered high-priority for 1.3 while the others are considered lower priority.

(BTW I disagree with @mqliang that the difference is about what granularity of object they apply to. All of them apply to pods.)

For forgiveness we can start with a very simple initial version that just takes a spec of how long the pod should to stay bound to a node before NodeController will evict it. Forgiveness in this case does not require any new persistent state (other than the spec of the requested forgiveness duration itself) because the last-update-time of the NodeStatus (or NodeCondition or whatever it is) already provides the timestamp against NodeController would compare "now+forgiveness_duration" to determine whether to evict. There is no common objects or implementation to share with DisruptionBudget in this case.

The initial implementation of tolerations will not have forgiveness nor will it include any runtime eviction (whether due to labels changing or a new taint being added). So before we add forgiveness to tolerations, we need to move all the node problem type stuff over to it, and add runtime eviction to taints. There's a relatively small chance we could get to those two for 1.3, but no chance we could also get to adding forgiveness to tolerations. So we have plenty of time to debate how to represent forgiveness in tolerations. Personally, I think all we need is a time period (same as I suggested in the previous paragraph for the simple version of toleration). I don't see a compelling use case for a rate-based forgiveness (though I know people have proposed it in the past). In that case there would again be no common objects or implementation to share with DisruptionBudget.

mqliang commented:
@davidopp 
> I think we should definitely not try to implement forgiveness, tolerations, and DisruptionBudget at the same time.

Yes, let's focus on the urgent one first. But it's worthing thinking about forgiveness/tolerations if they have any common objects or implementation to share, so that we can implement forgiveness/toleration in the future without changing any semantics on disruptionbudget.

davidopp commented on pkg/api/types.go:
```diff
@@ -1407,6 +1411,69 @@ type ReplicationControllerList struct {
+// DisruptionToken is used to limit the rate of pod disruption.
+// When disruption (preemption and other evictions) happens, we create a DisruptionToken
+// (with DeletionTimeStamp, so that API server could delete it when it expired).
+type DisruptionToken struct {
```
Token usually represents opportunity to do something (e.g. opportunity to disrupt), not the disruption itself. So maybe just call this DisruptionInfo. But as I mentioned later I'm not sure you need a separate object for this at all; it seems you just need an array of timestamps in DisruptionStatus.

bgrant0607 commented:
Re. forgiveness, tolerations, and disruption budget: I'm not suggesting we need to implement them all now, but we need to explain how they do / do not overlap before any of these features graduate from alpha.

I'm pretty sure tolerations and forgiveness can be unified, and disruption budget looks very similar, except that it applies to multiple instances rather than individual instances.

mqliang commented:
@davidopp @bgrant0607 Addressed some comments.

TBD:
1) remove `MinReadySeconds`, since it is not necessary in the first version of DisruptionBudget ?
2) find a way to not use "Spec" or "Status" in DisruptionBudget, since it is not an API resource. I propose to: 
   * remove DisruptionBudget
   * rename DisruptionBudgetSpec as DisruptionBudgetStrategy, and put it(a pointer) in Service and Job(put it in Pod annotation in the first version). 
   * put DisruptionBudgetStatus(a pointer), in ServiceStatus and JobStatus(also put it in Pod annotation in the first version). 

And I have one question: @davidopp suggest to use ***Pod*** annotation in the first version, but I stil can not understand it. I think we should use ***Job/Service*** annotation, since DisruptionBudget applies to a collection of Pods, instead of a single pod, right? Is there any reason we should put it in pod annotation?

kargakis commented:
Why is DisruptionBudget added in Services and not in Deployments where we already use minAvailable? Is there a proposal for this somewhere?

mqliang commented:
@kargakis see https://github.com/kubernetes/kubernetes/issues/12611

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits []RateLimitSpec `json:"rateLimits,omitempty"`
```
Please call this RateLimit, not RateLimitSpec. RateLimitSpec will make people think it is the Spec of some API object called RateLimit, but such an API object does not exist.

davidopp replied:
Or RateLimitItem

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,omitempty"`
```
Since MinAvailable and MinReadySeconds are used together, maybe it would be better to group them. For example

```
type DisruptionBudgetSpec struct {
   RateLimits []RateLimit
   MinInstances MinInstances
}

type MinInstances struct {
   MinAvailable intstr.IntOrString
   MinReadySeconds int
}
```

davidopp replied:
I think using IntOrString is probably not a good idea. Do we have fractions or percentages anywhere else in the API? We have things like millicores but those are integers, not fractions. If we have fractions or percentages somewhere else, use that pattern, otherwise we should use the millicore pattern, like

// exactly one is set
MinAvailableNumberOfInstances *int
MinAvailablePercentageOfInstances *int

mqliang replied:
How about implement a `Percent` type in util/percent/percent.go? It's more straightforward and not difficult.

mqliang replied:
> I think using IntOrString is probably not a good idea.

Why? It seems that
```
type MinInstances struct {
   MinAvailable intstr.IntOrString
   MinReadySeconds int
}
```
and
```
 type MinInstances struct {
   MinAvailableNumberOfInstances *int
   MinAvailablePercentageOfInstances *int
   MinReadySeconds int
}
```

are equivalent. Or I have misunderstand your comments?

mqliang replied:
If you think `IntOrString` is too general, I think maybe we could implement a `IntOrPrecent` type.

davidopp replied:
Thanks for the pointer to RollingUpdateDeployment. Since we are already using this pattern, what you have is fine.

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+// RateLimitSpec describes the rate limit on disruptions (preemption and other evictions) across the corresponding
+// collection of pods
+type RateLimitSpec struct {
+	// BucketSize is the maximum number of DisruptionToken could exist during a period of time.
```
It looks like you removed DisruptionToken from the PR but forgot to update the comment?

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+	// BucketSize is the maximum number of DisruptionToken could exist during a period of time.
+	// Once there are more than BucketSize number of DisruptionToken during a period of time, we can
+	// not disrupt pods in a further step.
+	BucketSize inf.Dec `json:"bucketSize,omitempty"`
```
I think the bucket thing is going to confuse people. A more straightforward name would be NumAllowedDisruptionsPerDuration.

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+	// Once there are more than BucketSize number of DisruptionToken during a period of time, we can
+	// not disrupt pods in a further step.
+	BucketSize inf.Dec `json:"bucketSize,omitempty"`
+	// Duration defines how long a token lives in the bucket, from insertion until deletion
```
The bucket thing is an implementation detail. Just say "the time period over which NumAllowedDisruptionsPerDuration applies."

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// The current disruption rate that the collection of pods is experiencing.
+	// This is an array of DeletionTimestamp associated with DisruptionToken: insertion timestamp + RateLimitSpec.Duration.
```
Update comment; there's no more DisruptionToken

mml commented:
@mqliang a disruption budget applies to any set of pods that all refer to that disruption budget.  They may be managed by a replication controller, job, deployment, or daemonset... or something else.  This approach allows any set of pods to share a disruption budget without forcing us to specify how those pods are managed in the DisruptionBudget type.

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits   []RateLimit `json:"rateLimits,omitempty"`
```
Add to comment "If there are multiple entries in the list, all must be satisfied for an eviction to be allowed."

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits   []RateLimit `json:"rateLimits,omitempty"`
+
+	MinInstances MinInstances `json:"minInstances"`
```
I think this should be "omitempty".  If it is empty (and RateLimits is empty), then the system does not provide any eviction protection to your pod, just like today.

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,,omitempty"`
+	// Minimum number of seconds for which a newly created pod should be ready
```
s/should/must/
Also, a pod being "ready" has an official definition, so you don't need to say "without any of its container crashing"

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,,omitempty"`
```
I think you can remove "omitempty" here. If you make DisruptionBudgetSpec.MinStances be "omitempty" then there is no need for this to be "omitempty" and thus also no reason to have a default. (Also, you have two commas.)

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+// collection of pods
+type RateLimit struct {
+	// NumAllowedDisruptionsPerDuration is the maximum number of disruptions are allowed during a period of time.
+	NumAllowedDisruptionsPerDuration int `json:"numAllowedDisruptionsPerDuration,omitempty"`
```
Remove "omitempty"

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+	// NumAllowedDisruptionsPerDuration is the maximum number of disruptions are allowed during a period of time.
+	NumAllowedDisruptionsPerDuration int `json:"numAllowedDisruptionsPerDuration,omitempty"`
+	// The time period over which NumAllowedDisruptionsPerDuration applies
+	Duration time.Duration `json:"duration,omitempty"`
```
Remove "omitempty"

davidopp commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,57 @@ type ReplicationControllerList struct {
+// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// The current disruption rate that the collection of pods is experiencing.
```
Actually I guess this list should just be insertion timestamp. The thing you have now works fine if there is only one RateLimit, but if there are multiple RateLimits then you would need multiple arrays here. I don't remember anymore why I suggested to make these entries be insertion timestamp + RateLimit.Duration. I think just insertion timestamp is better. So I would change this to
```
// Array of timestamps of recent disruptions corresponding to this DisruptionBudget.
DisruptionInfo []unversioned.Time `json:disruptionInfo,omitempty"
```

davidopp replied:
Maybe add to the comment "A timestamp is removed from the list when its insertion time is farther in the past than the maximum RateLimit.Duration"

mqliang committed:
API definitation of DisruptionBudget

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
 }
 
+// DisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type DisruptionBudget struct {
```
My preference would be to make this an API resource and move it to extensions. That will reduce risk when introducing such a major new mechanism.

We'd need to add a selector to identify the pods to which it applied.

That would make it easier to orphan and adopt pods, using the same mechanism as controllers and services.

I think it is ok and probably even desirable for multiple disruption budgets to be applied to a pod simultaneously, such as when a pod is a member of a canary or shard and also a member of a whole application tier.

davidopp replied:
I think the intention was to make it an API resource.

But I'm not sure I agree about adding a selector. I don't think this is an object normal users should be creating or thinking about. IMO it should be created by the controller managing a Deployment and the controller managing a Job (we'd extend the spec for those to include specifying a disruption budget), which should also be responsible for maintaining the back-pointers from the pods they manage to the DisruptionBudget object. Given the back-pointers, I don't think a selector on DisruptionBudget is needed. This trades off some flexibility but I think there are significant usability benefits to saying a DisruptionBudget maps 1:1 with those controllers. (In this model there is also just one DisruptionBudget per pod.)

A slightly more complicated model is to associate DisruptionBudget with Service (instead of Deployment or Job). In that case a pod can be controlled by more than one DisruptionBudget. But you still wouldn't need a selector in the DisruptionBudget.

bgrant0607 replied:
It's going to be tedious to keep a backpointer up to date.

Examples:
* Remove or change a label to orphan a pod for debugging.
* Delete a controller and orphan its pods.
* Adopt a pod under a new controller.
* Create a new service for a set of pods.

I'll try to think more about other alternatives.

bgrant0607 replied:
Re. API object: Needs TypeMeta and ObjectMeta.

And this should be moved out of the v1 API.

mml replied:
> It's going to be tedious to keep a backpointer up to date.
> 
> Examples:
> 
> Remove or change a label to orphan a pod for debugging.
> Delete a controller and orphan its pods.
> Adopt a pod under a new controller.
> Create a new service for a set of pods.

I understand these reasons, but using a selector has its own problems.  I would rather avoid defining semantics for multiple DBs applying to the same pod, and implementation becomes more complicated, too.  A pod *could* point to an arbitrary number of DBs, but without transactions or some other form of serialization, we can't guarantee that we've gotten approval from all the DBs and also charged it against all of them.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	MinInstances MinInstances `json:"minInstances,omitempty"`
+}
+
+type MinInstances struct {
```
I would prefer to eliminate the additional level of nesting here. Just inline the field(s) into DisruptionBudgetSpec.

davidopp replied:
I suggested the extra level of nesting. The benefit is that it makes it clearer that there are two kinds of disruption budgets being specified, a rate one and a shard-strength one. With three fields it requires a bit of reading (I'll admit not much) to realize that two of the fields together specify one concept.

bgrant0607 replied:
Unless we expect to add lots more fields, I'd prefer to avoid the additional nesting, but it's a weak preference.

xiang90 replied:
Is there a reason we called this MinInstances instead of MinPods?

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+type DisruptionBudgetStatus struct {
+	// Array of timestamps of recent disruptions corresponding to this DisruptionBudget.
+	// A timestamp is removed from the list when it is farther in the past than the maximum RateLimit.Duration.
+	DisruptionInfo []unversioned.Time `json:"disruptionInfo,omitempty"`
```
1. Field name: "info" isn't helpful. How about "recentDisruptions"?

2. We need to structure this in a way that we can bound it in size without causing blatantly incorrect behavior. A single disruption budget might cover tens of thousands of pods. It needs to have some kind of buckets. 

3. I also imagine we might distinguish between different types of disruptions in the future, especially if we try to unify this with tolerations/forgiveness, so we'd need somewhere to insert the type. If we introduce buckets that would provide a natural place to insert the type.

bgrant0607 replied:
When you update API naming conventions to forbid "num", forbid "info" also.

davidopp replied:
> We need to structure this in a way that we can bound it in size without causing blatantly incorrect behavior. A single disruption budget might cover tens of thousands of pods. It needs to have some kind of buckets.

Yeah, I was thinking about this but didn't have any good ideas yet. It's obviously a problem. How precise do we want the rate limit to be? I think the current approach (a list of every disruption, where a disruption is inserted when it happens and is removed after it has been in the list for "period" amount of time) is required if we want arbitrary granularity and ability to faithfully enforce SLOs across control plane failures/restarts (as opposed to making them best effort -- though I guess we could say rates are best-effort and shard-strength is guaranteed, which makes this a lot simpler). Batching could help but I think that also weakens the guarantee, just not by as much (e.g. if the component that is doing the batching crashes and loses an uncommitted batch).

bgrant0607 replied:
etcd has a hard size limit, and with just one exception, which we're going to have to fix soon, I've tried to ensure we don't put unbounded-size lists into individual objects/keys.

An example use case might help.

* 1 disruption per 10 minutes
* 10 disruptions per day

I don't think we want to take the longest period, a day in this case, and divide it into N (say, 100) buckets, since that would lose too much accuracy for smaller periods. We probably need N buckets per time period, where N is on the order of 10 or 20. We should likely omit buckets with 0 events. We could identify the buckets by number (0 to N-1) in order to avoid repeating timestamps in each bucket. We could explicitly represent the number of buckets to allow us to change that in the future.

Something like:

An array of:
* period length
* number of buckets
* array of {bucket index, disruption count}

And one timestamp indicating the time of the last update of all the disruption counts.

bgrant0607 replied:
It's also worth looking at the similar work done in the metrics APIs.

bgrant0607 replied:
I take it back -- metrics are a bit different because they just report aggregate values for each time period.

davidopp replied:
Just to make sure I understand -- with bucketing, you map all the disruptions that happen during some time period to a single time. So for example if the disruption period is one week and you have 7 buckets, you treat all of the disruptions that happen on Monday as if they happened at 00:00:00 on Monday, all the disruptions that happen on Tuesday as if they happened at 00:00:00 on Tuesday, and so on. So at 00:00:00 on each day, you clear the old count in that bucket, and start counting into that day's bucket?

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+// collection of pods
+type RateLimit struct {
+	// NumAllowedDisruptionsPerDuration is the maximum number of disruptions are allowed during a period of time.
+	NumAllowedDisruptionsPerDuration int `json:"numAllowedDisruptionsPerDuration"`
```
We don't generally use "num" in the API. It doesn't add anything.

bgrant0607 replied:
Please add this to naming conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions

bgrant0607 replied:
I'm not convinced that "PerDuration" is necessary, but if we keep it, it needs to be changed to "PerPeriod", as mentioned below.

xiang90 replied:
This is a max, but we do not have max perfix. For the previous min one we have min perfix. 

/cc @bgrant0607  do we have a rule for adding min/max prefix? Can we be consistent, at least for the Disruption thing here?

xiang90 replied:
also my question is why per second is not enough? we do not allow float64 rate?

bgrant0607 replied:
max is fine.

no floats.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	// NumAllowedDisruptionsPerDuration is the maximum number of disruptions are allowed during a period of time.
+	NumAllowedDisruptionsPerDuration int `json:"numAllowedDisruptionsPerDuration"`
+	// The time period over which NumAllowedDisruptionsPerDuration applies
+	Duration time.Duration `json:"duration"`
```
We don't use the term "Duration" in the API anywhere. We use "Period". This is actually already documented:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions

bgrant0607 replied:
Also should be unversioned.Duration.

bgrant0607 replied:
I realize that minReadySeconds was in terms of seconds in order to be consistent with Deployment, but that trades of intra-API consistency for inter-API consistency.

We could potentially also use seconds here. We haven't used unversioned.Duration anywhere in the actual API yet that I'm aware of -- just in componentconfig. It is more human-friendly than seconds, but harder for non-Go clients to decode.

smarterclayton replied:
If the scale is lower bounded by a second I think using seconds as the unit is easier for callers to reason about.  Reserving duration or finer grained intervals for future API use is better.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// Array of timestamps of recent disruptions corresponding to this DisruptionBudget.
+	// A timestamp is removed from the list when it is farther in the past than the maximum RateLimit.Duration.
```
When these comments are copied into the versioned API file, they'll be used to generate user-facing documentation. Therefore, they shouldn't mention Go fields or types. References should be in terms of the json field names.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	MinAvailable intstr.IntOrString `json:"minAvailable"`
+	// Minimum number of seconds for which a newly created pod must be ready for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready).
+	MinReadySeconds int `json:"minReadySeconds"`
```
Should allow a default value, so make this optional (omitempty).

xiang90 replied:
I do not feel MinReadySeconds is very descriptive. This is under `MinInstances` struct, so intuitively users assume this field describes MinInstances, but this actually defines `Available` (what do we mean Available for the first field in the struct). So I feel this is confusing.

mml replied:
As I'm now trying to minimize the spec here so we can easily implement the first version, I think we can leave MinReadySeconds out.  The Readiness probe already provides a great deal of customization around this area, and I think it's rich enough by itself.

bgrant0607 commented:
FYI, there is a proposal for cascading deletion: #23656

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	UnavailableReplicas int `json:"unavailableReplicas,omitempty"`
+}
+
+// RateLimit describes the rate limit on disruptions (preemption and other evictions) across the corresponding
```
We either need to scale all disruption counts by the collection size implicitly, or allow either an absolute number or percentage to be specified, as in minAvailable. For instance, 10% of the pods in the set per day.

bgrant0607 commented:
The priority proposal is strongly related: #22217

mml commented:
@mqliang the original goal of the PR was to unblock other projects that might cause disruptions.  Those need two things: (1) a way to record disruptions, and (2) a way to check if those disruptions are OK.

For (1) I suggest it's not the job of the component causing the disruption to record it directly, but rather the API server would update the relevant object during each DELETE.

For (2), we haven't fully designed it yet, but the sketch I have is that some controller manager thread ("disruption manager") examines the API objects representing the budgets, the PODs, and the disruptions, and then toggles a boolean flag on each budget that simply reflects "one more disruption's allowed" or not.  The API server would, once it's ready to enforce the budgets, check and then clear this bit when it allows a DELETE, and record the disruption as normal.  If the bit is not set when a DELETE comes in, it returns HTTP 409 or some other error code.

With this design, other components don't need to understand the disruption budget's type or its implementation.  They just need to deal properly the error code they get when a disruption isn't approved.

If we're on the same page, I'd like to suggest designing a smaller type that will be easier to implement.  We can expand it later.

My suggestion is this:  If we narrow our scope to MinAvailable, we won't have to consider time or history.  In fact, this version is so easy to implement, we don't really need an external process to manage evaluation.  The API server, when asked for a DELETE, can simply count up the available pods and return yes or no.

WDYT?

mqliang commented:
@mml
>  we haven't fully designed it yet, but the sketch I have is that some controller manager thread ("disruption manager") examines the API objects representing the budgets, the PODs, and the disruptions, and then toggles a boolean flag on each budget that simply reflects "one more disruption's allowed" or not. The API server would, once it's ready to enforce the budgets, check and then clear this bit when it allows a DELETE, and record the disruption as normal. If the bit is not set when a DELETE comes in, it returns HTTP 409 or some other error code.
>
With this design, other components don't need to understand the disruption budget's type or its implementation. They just need to deal properly with the error code they get when a disruption isn't approved.

Such a implementation SGTM. ResourceQuota also implement like this: a "resourcequota manager" in controller-manager examine the resourcequota status,  and a '"resourcequota admission controller" in API server just check the status and allow/disallow a request. 

> If we're on the same page, I'd like to suggest designing a smaller type that will be easier to implement. We can expand it later.
>
My suggestion is this: If we narrow our scope to MinAvailable, we won't have to consider time or history. In fact, this version is so easy to implement, we don't really need an external process to manage evaluation. The API server, when asked for a DELETE, can simply count up the available pods and return yes or no.

Sounds good to me.

davidopp commented:
> For (1) I suggest it's not the job of the component causing the disruption to record it directly, but rather the API server would update the relevant object during each DELETE.

I agree. The design doc actually suggested two phases, one where Status is just based on observations, and then a second phase where "the API server explicitly debit[s] the DisruptionBudgetStatus when it accepts an /evict" but I think the first phase was a bad idea. But in any event, I don't think anything in the current version of the API makes it unsuitable for being updated by the API server, correct?

> For (2), we haven't fully designed it yet, but the sketch I have is that some controller manager thread ("disruption manager") examines the API objects representing the budgets, the PODs, and the disruptions, and then toggles a boolean flag on each budget that simply reflects "one more disruption's allowed" or not. The API server would, once it's ready to enforce the budgets, check and then clear this bit when it allows a DELETE, and record the disruption as normal. If the bit is not set when a DELETE comes in, it returns HTTP 409 or some other error code.

This seems like a good idea (API server wouldn't need to understand any of the fields of DisruptionBudget except this bool). But we need to make sure there aren't any race conditions and that it works across API server and controller restart and with multiple API servers. The reason I'm a little concerned about races is that etcd v2 doesn't have multi-object transactions, so the disruption record (DisruptionInfo) and the bool you're proposing can't be updated atomically. I think it's OK as long as API server always updates disruption record before updating bool and API server only flips the bool from true to false, but I haven't thought it through carefully. The one downside (assuming it's not racy) is that (if I'm understanding your proposal correctly) it requires a round-trip to the controller to reset the bit after each eviction. This may or may not be a problem depending on the rate of evictions and the number of pods covered by a typical DisruptionBudget.

xiang90 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	// A timestamp is removed from the list when it is farther in the past than the maximum RateLimit.Duration.
+	DisruptionInfo []unversioned.Time `json:"disruptionInfo,omitempty"`
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
+	AvailableReplicas int `json:"availableReplicas,omitempty"`
```
What is the difference between Replicas and Instances? Why do we call this replicas here?

bgrant0607 replied:
I think that's because it was copied from Deployment.

xiang90 commented on pkg/api/types.go:
```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
+	AvailableReplicas int `json:"availableReplicas,omitempty"`
+	// Total number of unavailable pods targeted by this disruption budget.
+	UnavailableReplicas int `json:"unavailableReplicas,omitempty"`
```
same question here. why replicas?

mml commented:
Something like https://github.com/kubernetes/kubernetes/compare/master...mml:disruption-budget ?

davidopp commented:
@erictune made a good point today, that no matter how we decide to associate pods with DisruptionBudgets (back-pointer, label selector, etc.) we need (perhaps not in the first version, but eventually) a way to prevent users from changing that mapping in a way that violates policy. The design doc is a bit hand-wavy about the policy; it just says

> Initially, disruption budgets will be specified by the user. Thus as with priority, disruption budgets need to be tied into quota, to prevent users from saying none of their pods can every be disrupted. The exact way of expressing and enforcing this quota is TBD, though a simple starting point would be to assign a default disruption budget based on priority level (more liberal with decreasing priority). 

but @erictune's point is that you don't want users to be able to arbitrarily block all of their pods from ever being evicted. This can be enforced with an admission controller that passes judgement on changes to the back-pointer (if we use the back-pointer approach) or changes to the pod's labels (if we use the label selector approach). It definitely seems that the back-pointer approach is simpler in this respect (for example, you could start with a trivial policy that DisruptionBudget can only be set at pod creation time and can never be updated; such a policy obviously wouldn't work in the label approach). 

We don't need anything like this for the first version, but we should at least make sure whatever we do for the first version can reasonably have policies like this applied to it in the future.

kargakis commented:
cc: @ironcladlou @smarterclayton

davidopp commented:
Oh, also there was some discussion of whether DisruptionBudget should be namespaced. I think that's fine, as I can't think of any reason you'd need the covered pods to come from multiple namespaces. So it makes sense to just put the DisruptionBudget in the same namespace as the pods it covers.

davidopp commented:
Also we discussed that if we use the back-pointer approach, should the back-pointer be an ObjectRef or just the name of a DisruptionBudget. I guess we were leaning towards name because otherwise the reference would have to cross API group (e.g. from v1 (for pod) to disruption/v1alpha1 (for DisruptionBudget)).

smarterclayton commented:
Agreed it should be namespaced.

bgrant0607 commented:
I like the idea of pruning back to just minAvailable.

Just name for backrefs also SGTM, especially since it should be specified via annotation initially.

Some background thoughts:

I think we all agree that requiring users to create and delete another resource for every service or controller is not ideal. I'm also somewhat concerned that it won't interact well with whatever preemption scheme we eventually develop, and that it may be too operationally complex to reason about. It would be easier to think in terms of availability classes than individual counters, though the latter will certainly need to be part of the implementation.

The main reasons a per-collection resource is needed are:
1. To identify the set (at least a name for the distinct budget, if not a selector as well)
2. To serve budget status for the set

Assuming we do need per-collection resources, it would be nice to be able to auto-generate DisruptionBudget in a way that was configurable by admins. We've been developing a model for policy objects, like LimitRange, MetadataPolicy, PodSecurityPolicy, etc. A similar sort of approach would be nice.

A way to resolve (1) would be to provide a "formula" for how to map from a pod to a budget. It wouldn't provide the full power of our more general label selectors, but a selector (to match pods upon which the generator would be applied) plus a list of label keys could act as such a formula: each unique set of values for those keys would map to a distinct budget. 

One problem with performing this mapping on a per-pod basis is how to GC unneeded DisruptionBudget objects. We wouldn't want to add every pod as an individual parent reference to the DB object. We'd also prefer not to develop yet another GC mechanism (e.g., TTL-based). However, linking a DB object to controllers wouldn't be extensible to user-developed controllers (except maybe ones implemented using federated apiservers).

mml commented:
> 2. To serve budget status for the set

I am not convinced that the budget status needs to be part of the API type we expose to users or even cached in etcd.  If we start with minAvailable, it'll be trivial to evaluate and we don't need to keep state, but even in other cases I think that we could simply find a compact way to encode the disruption events with enough information that the DB evaluator can relatively quickly give a correct go/no-go response, even involving events before the DB itself was created.

That encoding would go in etcd at first, but ultimately it's a lot like recording timeseries data for monitoring or a transaction log that we have particular query patterns on, and we want the flexibility to change where and how we store it.

I think it's desirable to omit the state from the public API, and I think we [I] can design something (using an object we store in etcd but don't expose to users) by the time we're ready to implement rate limits.

mml commented:
> One problem with performing this mapping on a per-pod basis is how to GC unneeded DisruptionBudget objects.

The same loop that evaluates the formula could presumably also identify unneeded DBs.  It's already comparing the desired set of DBs with the actual set.  With a little hysteresis added, it can the delete the unneeded ones.  This amounts to another GC mechanism, but it's tightly integrated with the system generating the objects in the first place.

davidopp commented:
I think just doing MinAvailable for the first version is fine. The design doc actually suggested doing a stateless version before the stateful one, but I was hoping we had a good enough handle on how to do the stateful one that it wasn't necessary. I actually think we've kinda converged on how to do the stateful version, modulo the question of whether the record of disruptions is public or private state. AFAIK we don't actually have any pattern for doing private controller state, though I don't object to the concept. But again, I'm fine with just doing MinAvailable for the first version, which is stateless.

I'm totally in favor of auto-generating DisruptionBudgets, though I think that in the first alpha version we should just require users to create and GC them manually. We can layer auto-generation and GC on top of that in the next iteration.

> The same loop that evaluates the formula could presumably also identify unneeded DBs. It's already comparing the desired set of DBs with the actual set.

I assume the loop you're talking about is the admission controller @bgrant0607 proposed that assigns a DisruptionBudget to a pod based on the pod's characteristics?

mml commented:
@davidopp I was referring to where @bgrant0607 says "formula" below.  I don't see mention of an admission controller.

> Assuming we do need per-collection resources, it would be nice to be able to auto-generate DisruptionBudget in a way that was configurable by admins. We've been developing a model for policy objects, like LimitRange, MetadataPolicy, PodSecurityPolicy, etc. A similar sort of approach would be nice.

> A way to resolve (1) would be to provide a "formula" for how to map from a pod to a budget. It wouldn't provide the full power of our more general label selectors, but a selector (to match pods upon which the generator would be applied) plus a list of label keys could act as such a formula: each unique set of values for those keys would map to a distinct budget.

davidopp commented:
> I was referring to where @bgrant0607 says "formula" below

I interpreted

> Just name for backrefs also SGTM

plus the part you quoted to imply that what he had in mind was that there would be a LimitRanger-style admission controller that writes the backref at pod creation time based on the formula/policy.

I definitely think we want this to be composable and incrementally deployable, i.e. implement the DB object and backref from pod to DB (and admission controller that makes the admit decision) first as the "lowest layer" mechanism, then implement the DB auto-generation and policy stuff on top.

mml commented:
> I definitely think we want this to be composable and incrementally deployable, i.e. implement the DB object and backref from pod to DB (and admission controller that makes the admit decision) first as the "lowest layer" mechanism, then implement the DB auto-generation and policy stuff on top.

Agreed.

bgrant0607 commented:
Sounds like a narrowly scoped initial iteration, which SGTM. 

Did discussions change any ideas/opinions about the implementation approach?

mml commented:
There seem to be three axes around implementation.

1. Should it be a synchronous or asynchronous API?
2. If synchronous, should the approval check happen in the apiserver?  Or should we stand up a separate server that implements the check?
3. If apiserver, should it be as an admission controller?  
4. Should there be an uncoupled control loop evaluating budgets, state, and history and recording its results, or should evaluation happen at the same place we make the check?

For (1) I strongly favor synchronous.  It's easier to implement and reason about and leaves less opportunity for errors on the client side.  You ask and get either yes or no.

For (2) I favor implementing it in the apiserver as a subresource of the pod, but if we do that we can't label it as alpha.  @lavalamp suggests creating a generic "evict" path in a "disruptions" API group and simply posting to it the pod we'd like to evict.  The shape of this mainly allows us to abide by our own API versioning rules and keep this labeled alpha, but it's not the design I'd choose if I didn't have those rules.  (Someday it's worth considering how often we end up doing this to ourselves.)  I don't want the complexity of standing up and deploying a separate server right now, but we should consider if we want to as the implementation evolves.

For (3) not an admission controller.  It seemed elegant, but I've been swayed.

For (4) I now think of that as a future optimization.  Our first implementation is so simple that we can evaluate the rules immediately and give a result right in the handler.

lavalamp commented:
(as discussed in person-- subresources are versioned differently than their parents, so that should not be a factor)

mml commented:
Summarizing from a recent discussion where these decisions differ from my last update...

1. Due to the complications of auth, this will be easiest to implement if it's just a DELETE on a pod with a special field (query param or delete option) indicating that it's supposed to be a within-budget eviction.
2. That said, if we find out that implementations of this field are silently ignored where not implemented, we'll need a different plan, possibly a subresource.
3. To limit the amount of work done within the api server, we will have a separate controller that evaluates budgets and system state and communicates with a bit indicating whether a DELETE is approved or not.  The DELETE handling code then checks this bit, and only if it's set will it clear the bit then approve the DELETE.
4. To begin with, we'll also toggle the feature with a flag.

mml commented:
As predicted, deleting with either an unknown query parameter or a nonsense field is just silently ignored.

Here's how I tested.  In both cases I got no error message and the pod was gone.  Looks like we will need a subresource after all, otherwise clients will have no way of knowing if they are getting the guarded evictions they're asking for.

```console
% curl --insecure -u admin:$p -X DELETE -H 'Content-type: application/json' --data-binary '{ "foo": true }' https://104.197.207.134/api/v1/namespaces/default/pods/foobar-3073401103-zqpah
% curl --insecure -u admin:$p -X DELETE 'https://104.197.207.134/api/v1/namespaces/default/pods/foobar-3073401103-zwqjn?foo=bar'
```

davidopp commented:
One thing I forgot: We want the evict request to be accompanied by an optional parameter that indicates whether the controller responsible for the pod should avoid placing pods on that node (more specifically, whether the API server should add the pod's controller to the PreferAvoidPods field of the node if it does the eviction -- see #20699 and the design doc).

lavalamp commented:
I'm beginning to get somewhat uncomfortable with the amount of work we're
going to have to do in the DELETE handler. I guess as long as it's still
constant time it's acceptable.

On Mon, Apr 18, 2016 at 12:37 AM, David Oppenheimer <
notifications@github.com> wrote:

> One thing I forgot: We want the evict request to be accompanied by an
> optional parameter that indicates whether the controller responsible for
> the pod should avoid placing pods on that node (more specifically, whether
> the API server should add the pod's controller to the PreferAvoidPods field
> of the node if it does the eviction -- see #20699
> <https://github.com/kubernetes/kubernetes/pull/20699> and the design doc).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22774#issuecomment-211244679>
>

mml commented:
It is only constant time if we've already added `controllerRef` to pod, which I guess is now a blocker for the `PreferAvoidPods` subfeature.  I guess that's fine since it's already on the list for 1.3.

ihmccreery commented:
> One thing I forgot: We want the evict request to be accompanied by an optional parameter that indicates whether the controller responsible for the pod should avoid placing pods on that node

I know there are other use-cases besides node deletion, but for node deletion, we shouldn't need `PreferAvoidPods`, since the nodes will presumably be tainted as unschedulable anyway.  So for upgrades, I'm more than happy with an implementation of this before `PreferAvoidPods` exists.

ihmccreery commented:
And, a UX-type scenario:

I want to take a node offline (I'm a sysadmin doing hardware maintenance on a bare-metal machine, or I'm an automated upgrade controller trying to recreate cloud instances at a new version).  I want to decide if I can take down `node-1`.  How does this API help me do that?  I can ask about relocating pods that have a `DisruptionBudget`, but what about those that don't?  Are we going to have sane defaults?  If none of my pods have `DisruptionBudget`s, do I have to be intelligent and not ask to take down everything at once?

mml commented:
@ihmccreery the feature is only ready to support your use case when it's mature enough that we can require users to have DBs for every important pod.  My guess is that will hinge upon some way to stamp out new DB objects like @bgrant0607 was describing above.

The narrower version currently described doesn't make it easy to guarantee that every important pod has a DB.

ihmccreery commented:
> the feature is only ready to support your use case when it's mature enough that we can require users to have DBs for every important pod

What use case(s) would *not* require every important pod to have a `DisruptionBudget`?

mml commented:
Getting anything into the hands of implementors of reschedulers, drainers, etc.  Clusters where the admin is willing to manually manage the DBs while this is in alpha.

It's clear this isn't the full design we'd use when pushing the API out of alpha, but it does unblock other work and allows us to get feedback and experience.

mqliang commented:
@mml I will fly to Beijing tonight for a conference, and will have a full agenda in the next few days. Since DisruptionBudget is really urgent, would mind pick this PR up? If you'd like to help, feel free to close this and open a new one.

mqliang commented:
@mml I may could find some time nex week to take a look at your PR(or update this).

mml commented:
No problem, @mqliang.  I will copy your branch and start a fresh PR.  Thanks for your work!

mqliang commented:
close this in favour of https://github.com/kubernetes/kubernetes/pull/24697
